### PR TITLE
Parallel MambaFF

### PR DESCRIPTION
**Pull request by Krystian-Krol**
- Fix double residual
- Parallelise Mamba + FF
- Layer norm for each layer

# Description
- Fixed bug where parallel was doing residual stream on top of another residual stream
- Added and tested parallel implementation, now both versions have similar number of parameters in side-by-side view

[Neptune link](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9ac46a51-c84e-442d-963c-ac6a3a8af3f3&detailsTab=charts&shortId=LLMRANDOM-7224&type=run)
# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/train/cc_train.py [hunk 1]
```diff
@@ -159,6 +159,12 @@ def main(
         else:
             raise ValueError(f"Unknown module name: {module_name}")
 
+    if args.parallel_blocks:
+        modules = block_modules.values()
+        block_modules = {
+            "parallel": lambda: Parallel(*[module() for module in modules])
```
Comment (sebastianjaszczur) [inline research/conditional/train/cc_train.py:165]: I think this makes both blocks share the same layernorm. Maybe it's good, maybe it isn't; probably doesn't matter too much. But it's at least worth noting.

Comment (sebastianjaszczur) [reply research/conditional/train/cc_train.py:165]: Intuitively, I would like both of those blocks to have separate layernorms, but I'm not sure.

File: research/conditional/train/configs/experiments/expert_choice/mamba_parallel.yaml
Comment (maciek-pioro) [inline research/conditional/train/configs/experiments/expert_choice/mamba_parallel.yaml:1]: Can we get rid of that?

File: research/conditional/utils/model_utils.py [hunk 2]
```diff
@@ -219,14 +219,26 @@ def get_attention_layer(args):
...
-        raise NotImplementedError(f"Norm type {args.norm_class} not implemented")
+        raise NotImplementedError(f"Norm type {norm_class} not implemented")
+
+
+def get_residual_layer(args):
+    norm_class = get_norm_class(args.norm_class)
     if args.residual_mode == "pre_norm":
+        if args.parallel_blocks:
+            return partial(
+                llm.PreNormBlock, dmodel=args.dmodel, norm_class=torch.nn.Identity
```
Comment (maciek-pioro) [inline research/conditional/utils/model_utils.py:236]: Can we make e.g. `llm.ParallelPreNormBlock`? Now it looks like we have some part of the logic for creating a block in [`research/conditional/utils/model_utils.py`, `research/conditional/train/cc_train.py`] and another part in `lizrd/core/llm.py`, so basically it's split between `research` and `lizrd`, while actually it should reside in `lizrd`.

**Review by maciek-pioro — CHANGES_REQUESTED** [review]
Good job working on that! Can we make some adjustments to the structure of the code (info in the comment)?

**Review by maciek-pioro — APPROVED** [review]

**Comment by maciek-pioro** [comment]
Nice, lgtm

**Pull request merged**